cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+			// go through group by group
+			for _, cluster := range subclusters.UnsortedList() {
+				status := statusFunc(cluster)
+				status.GroupKey = key
+
+				newstatus, needtorollout := needToRollout(status, timeout)
+				if needtorollout {
+					rolloutclusters[cluster] = status
+				}
+				if newstatus == TimeOut {
+					status.Status = newstatus
+					timeoutclusters[cluster] = status
+				}
+			}
+
+			if len(rolloutclusters) > 0 {
+				return RolloutResult{
+					ClustersToRollout: rolloutclusters,
+					ClustersTimeOut:   timeoutclusters,
+				}
+			}
+		}
+	}
+
+	return RolloutResult{
+		ClustersToRollout: rolloutclusters,
+		ClustersTimeOut:   timeoutclusters,
+	}
+}
+
+// Check if the cluster need to update based on existing cluster status and timeout.
+// Timeout is used for status progressing and failed:
+// 1) When timeout is None (maxTimeDuration), it means will wait until reach success status.
+// Return true to append it to the result and stop rollout other clusters or groups.
+// 2) Timeout is 0 means continue upgrade others without any wait.
+// Return false to skip updating it and continue rollout other clusters or groups.
+func needToRollout(status ClusterRolloutStatus, timeout time.Duration) (RolloutStatus, bool) {
+	switch status.Status {
+	case ToApply:
+		return status.Status, true
+	case Progressing:
+		if beforeTimeOut(status.LastTransitionTime, timeout) {
+			return Progressing, true
+		}
+		return TimeOut, false
+	case Failed:
+		if beforeTimeOut(status.LastTransitionTime, timeout) {
+			return Failed, true
+		}
+		return TimeOut, false
+	case TimeOut:
+		return status.Status, false
+	case Succeed:
+		return status.Status, false
+	case Skip:
+		return status.Status, false
+	default:
+		return status.Status, true
+	}
+}
+
+// check if current time is before start time + timeout duration
+func beforeTimeOut(startTime *time.Time, timeout time.Duration) bool {
+	var timeoutTime time.Time
+	if startTime == nil {
+		timeoutTime = RolloutClock.Now().Add(timeout)
+	} else {
+		timeoutTime = startTime.Add(timeout)
+	}
+	if RolloutClock.Now().Before(timeoutTime) {
+		return true
+	}
+	return false
+}
+
+func calculateLength(maxConcurrency intstr.IntOrString, total int) (int, error) {
+	length := total
+
+	switch maxConcurrency.Type {
+	case intstr.Int:
+		length = maxConcurrency.IntValue()
+	case intstr.String:
+		str := maxConcurrency.StrVal
+		if strings.HasSuffix(str, "%") {
+			f, err := strconv.ParseFloat(str[:len(str)-1], 64)
+			if err != nil {
+				return length, err
+			}
+			length = int(math.Ceil(f / 100 * float64(total)))
+		} else {
+			return length, fmt.Errorf("%v invalid type: string is not a percentage", maxConcurrency)
+		}
+	default:
+		return length, fmt.Errorf("incorrect MaxConcurrency type %v", maxConcurrency.Type)
+	}
+
+	if length <= 0 || length > total {
+		length = total
+	}
+
+	return length, nil
+}
+
+func parseTimeout(timeoutStr string) (time.Duration, error) {
```

Comment:
Why not use `time.ParseDuration`?

Reply:
The unit are different, so not use it: 
- For `time.ParseDuration` valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h". 
- For our timeout, it follows pattern `"^(([0-9])+[h|m|s])|None$"`